lib/recipes-data/src/lib/extract-recipes.ts

```diff
@@ -0,0 +1,38 @@
+import type {Block} from "@guardian/content-api-models/v1/block";
+import type {Blocks} from "@guardian/content-api-models/v1/blocks";
+import type {Content} from "@guardian/content-api-models/v1/content";
+import {ContentType} from "@guardian/content-api-models/v1/contentType";
+import {ElementType} from "@guardian/content-api-models/v1/elementType";
+import type {RecipeReferenceWithoutChecksum} from './models';
+
+export function extractAllRecipesFromArticle(content: Content): RecipeReferenceWithoutChecksum[] {
+  let allRecipesInAnArticle = Array<RecipeReferenceWithoutChecksum>()
+
+  if (content.type == ContentType.ARTICLE && content.blocks) {
+    const articleBlocks: Blocks = content.blocks
+    const getAllMainBlockRecipesIfPresent: RecipeReferenceWithoutChecksum[] = extractRecipeData(content.id, articleBlocks.main as Block)
+    const bodyBlocks = articleBlocks.body as Block[]
+    const getAllBodyBlocksRecipesIfPresent: RecipeReferenceWithoutChecksum[] = bodyBlocks.flatMap(bodyBlock => extractRecipeData(content.id, bodyBlock))
+    allRecipesInAnArticle = getAllMainBlockRecipesIfPresent.concat(...getAllBodyBlocksRecipesIfPresent)
+  }
+  return allRecipesInAnArticle
+}
+
+export function extractRecipeData(canonicalId: string, block: Block): RecipeReferenceWithoutChecksum[] {
+  const allRecipes = block.elements
+    .filter(elem => elem.type === ElementType.RECIPE)
+    .map(recp => parseJsonBlob(canonicalId, recp.recipeTypeData?.recipeJson as string))
+  return allRecipes
+}
+
+function parseJsonBlob(canonicalId: string, recipeJson: string): RecipeReferenceWithoutChecksum {
+  const recipeData = JSON.parse(recipeJson) as Record<string, unknown>
+  if (!recipeData.id) {
+    throw new Error(`Error! No id present in the recipeJson, canonicalId is ${canonicalId}`)
```

Comment:
I was chatting with @tonytw1 about whether an exception is the most appropriate way to handle this case and we decided, on balance, that it's probably best to ignore invalid recipes and allow the rest through.

In practise, that means you'll need to change the `parseJsonBlob` function to return `RecipeReferenceWithoutChecksum | null` and return `null` in this case.  Please put a TODO comment in to indicate we should incorporate a metric for failed recipes so we can have an indication of upstream issues.

Once we are returning `RecipeReferenceWithoutChecksum | null` the method caller will get upset.... you'll need to put a `filter` stage in after the `map` that calls this function to filter out all `null` values.  If typescript doesn't realise that `null` is no longer possible then it's safe to cast the value using `as`.